File: src/main/java/org/asmolinski/java/Collections.java
package org.asmolinski.java;

import org.asmolinski.kotlin.CollectionsKt;

import java.util.List;
import java.util.Map;

import static java.util.stream.Collectors.partitioningBy;
import static java.util.stream.Collectors.toList;

public class Collections {

    public static void main(String[] args) {
        // map + filter
        System.out.println(squaresOfEvenInts(List.of(1, 2, 3, 4, 5, 6)));

        // partition
        final var partitioned = partitionOddVsEven(List.of(1, 2, 3, 4, 5, 6));
        System.out.println("Parzyste: " + partitioned.get(true));
        System.out.println("Nieparzyste: " + partitioned.get(false));

        // kotlin interoperability
        final var fromKotlin = CollectionsKt.squaresOfEvenInts(List.of(1, 2, 3, 4, 5, 6));
        System.out.println(CollectionsKt.squaresOfEvenInts(List.of(1, 2, 3, 4, 5, 6)));
        System.out.println("Suma " + fromKotlin.stream().mapToInt(it -> it).sum());
    }

    public static List<Integer> squaresOfEvenInts(List<Integer> input) {
        return input.stream()
                .filter(i -> i % 2 == 0)
                .map(i -> i * i)
                .collect(toList());
    }

    private static Map<Boolean, List<Integer>> partitionOddVsEven(List<Integer> input) {
        return input.stream()
                .collect(partitioningBy(i -> i % 2 == 0));
    }

}

File: src/main/kotlin/org/asmolinski/kotlin/Collections.kt
package org.asmolinski.kotlin

fun main() {
    // map + filter
    println(squaresOfEvenInts(listOf(1, 2, 3, 4, 5, 6)))

    // partition
    val (even, odd) = partitionOddVsEven(listOf(1, 2, 3, 4, 5, 6))
    println("Parzyste: $even")
    println("Nieparzyste: $odd")

    // java interoperability
    val fromJava = org.asmolinski.java.Collections.squaresOfEvenInts(listOf(1, 2, 3, 4, 5, 6))
    println(fromJava)
    println("Suma " + fromJava.sum())

    // sequence
    println(sequenceProcessing(listOf(1, 2, 3, 4, 5, 6)))
}

fun squaresOfEvenInts(ints: List<Int>) =
    ints
        .filter { it % 2 == 0 }
        .map { it * it }

fun partitionOddVsEven(ints: List<Int>) =
    ints.partition { it % 2 == 0 }

fun sequenceProcessing(ints: List<Int>) =
    ints.asSequence()
        .map { it * it }
        .mapIndexed { index, value ->  if (index % 2 == 0) value + 1 else value }
        .shuffled()
        .sum()